File: 719. Find K-th Smallest Pair Distance.go
func smallestDistancePair(nums []int, k int) int {
    sort.Ints(nums)
    n := len(nums)
    answer, left, right := 0, 0, nums[n-1] - nums[0]
	for left <= right {
		mid := (left + right) / 2
		count := getCount(nums, mid)
		if count < k {
			left = mid + 1
		} else {
			answer = mid
			right = mid - 1
		}
	}
	return answer
}

func getCount(nums []int, k int) int {
	answer, left := 0, 0
	for right := 0; right < len(nums); right++ {
		for nums[right] - nums[left] > k {
			left++
		}
		answer += right - left
	}
	return answer
}

File: java/719. Find K-th Smallest Pair Distance.java
class Solution {
    public int smallestDistancePair(int[] nums, int k) {
        int n = nums.length;
        Arrays.sort(nums);
        int left = 0;
        int right = nums[n-1]-nums[0];
        int answer = 0;
        while (left <= right) {
            int mid = (left + right) / 2;
            int count = getCount(nums, mid);
            if (count < k) {
                left = mid + 1;
            } else {
                answer = mid;
                right = mid - 1;
            }
        }
        return answer;
    }
    private int getCount(int[] nums, int k) {
        int left = 0;
        int answer = 0;
        for (int right = 0; right < nums.length; right++) {
            while (nums[right] - nums[left] > k) {
                left++;
            }
            answer += right - left;
        }
        return answer;
    }
}